drivers/service_upgrade.go

```diff
@@ -0,0 +1,223 @@
+	if !ok {
+		return http.StatusInternalServerError, fmt.Errorf("Can't process config")
+	}
+
+	if config.ServiceSelector == nil {
+		return http.StatusBadRequest, fmt.Errorf("Service selectors not provided")
+	}
+
+	if config.Image == "" {
+		return http.StatusBadRequest, fmt.Errorf("Image not provided")
+	}
+
+	if config.Tag == "" {
+		return http.StatusBadRequest, fmt.Errorf("Tag not provided")
+	}
+
+	if config.BatchSize <= 0 {
+		return http.StatusBadRequest, fmt.Errorf("Batch size for upgrade not provided/invalid")
+	}
+
+	if config.IntervalMillis <= 0 {
+		return http.StatusBadRequest, fmt.Errorf("Batch interval for upgrade not provided/invalid")
+	}
+
+	return http.StatusOK, nil
+}
+
+func (s *ServiceUpgradeDriver) Execute(conf interface{}, apiClient *client.RancherClient, requestPayload interface{}) (int, error) {
+	requestBody := make(map[string]interface{})
+	config := &model.ServiceUpgrade{}
+	err := mapstructure.Decode(conf, config)
+	if err != nil {
+		return http.StatusInternalServerError, errors.Wrap(err, "Couldn't unmarshal config")
+	}
+
+	image := config.Image
+	tag := config.Tag
+
+	if requestPayload == nil {
+		return http.StatusBadRequest, fmt.Errorf("No Payload recevied from Dockerhub webhook")
+	}
+
+	requestBody, ok := requestPayload.(map[string]interface{})
+	if !ok {
+		return http.StatusBadRequest, fmt.Errorf("Body should be of type map[string]interface{}")
+	}
+
+	pushedData, ok := requestBody["push_data"]
+	if !ok {
+		return http.StatusBadRequest, fmt.Errorf("Dockerhub webhook invalid data")
+	}
+
+	pushedTag, ok := pushedData.(map[string]interface{})["tag"].(string)
+	if !ok {
+		return http.StatusBadRequest, fmt.Errorf("Dockerhub webhook contains no tag")
+	}
+
+	repository, ok := requestBody["repository"]
+	if !ok {
+		return http.StatusBadRequest, fmt.Errorf("Dockerhub repository information not provided")
+	}
+
+	imageName, ok := repository.(map[string]interface{})["repo_name"].(string)
+	if !ok {
+		return http.StatusBadRequest, fmt.Errorf("Dockerhub repo name not provided")
+	}
+
+	pushedImage := imageName + ":" + pushedTag
+	requestedImage := image + ":" + tag
+	if requestedImage != pushedImage {
+		return http.StatusOK, nil
+	}
+
+	log.Infof("Image %s pushed in Docker Hub, upgrading services with serviceSelector %v", pushedImage, config.ServiceSelector)
+
+	go upgradeServices(apiClient, config, pushedImage)
+
+	return http.StatusOK, nil
+}
+
+func upgradeServices(apiClient *client.RancherClient, config *model.ServiceUpgrade, pushedImage string) {
+	var key, value string
+	serviceSelector := make(map[string]string)
+
+	for key, value = range config.ServiceSelector {
+		serviceSelector[key] = value
+	}
+	batchSize := config.BatchSize
+	intervalMillis := config.IntervalMillis
+	startFirst := config.StartFirst
+
+	services, err := apiClient.Service.List(&client.ListOpts{})
+	if err != nil {
+		log.Errorf("Error %v in listing services", err)
+		return
+	}
+
+	for _, service := range services.Data {
```

Comment:
@mrajashree after thinking about it more, I think we can fit in parallelizing these. You would just need to wrap some portion of the logic inside this for loop in a go routine. Does that make sense?